RecoCTPPS/TotemRPLocal/plugins/TotemRPUVPatternFinder.cc

```diff
@@ -0,0 +1,278 @@
+    double z0, double threshold_loc, unsigned int planes_required,
+    const DetSetVector<TotemRPRecHit> &hits, DetSet<TotemRPUVPattern> &patterns)
+{
+  // run recognition
+  DetSet<TotemRPUVPattern> newPatterns;
+  lrcgn->GetPatterns(hits, z0, threshold_loc, newPatterns);
+  
+  // set pattern properties and copy to the global pattern collection
+  for (auto &p : newPatterns)
+  {
+    p.setProjection(proj);
+
+    p.setFittable(true);
+
+    set<unsigned int> planes;
+    for (const auto &ds : p.getHits())
+        planes.insert(TotemRPDetId::rawToDecId(ds.detId()) % 10);
+
+    if (planes.size() < planes_required)
+      p.setFittable(false);
+    
+    if (fabs(p.getA()) > max_a_toFit)
+      p.setFittable(false);
+
+    patterns.push_back(p);
+  }
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void TotemRPUVPatternFinder::produce(edm::Event& event, const edm::EventSetup& es)
+{
+  if (verbosity > 5)
+    printf(">> TotemRPUVPatternFinder::produce (%u:%llu)\n", event.id().run(), event.id().event());
+
+  // geometry
+  ESHandle<TotemRPGeometry> geometry;
+  es.get<VeryForwardRealGeometryRecord>().get(geometry);
+  if (geometryWatcher.check(es))
+    lrcgn->ResetGeometry(geometry.product());
+  
+  // get input
+  edm::Handle< edm::DetSetVector<TotemRPRecHit> > input;
+  event.getByToken(detSetVectorTotemRPRecHitToken, input);
+
+  // prepare output
+  DetSetVector<TotemRPUVPattern> patternsVector;
+  
+  // split input per RP and per U/V projection
+  struct RPData
+  {
+    DetSetVector<TotemRPRecHit> hits_U, hits_V;
+    map<uint8_t, uint16_t> planeOccupancy_U, planeOccupancy_V;
+  };
+  map<unsigned int, RPData> rpData;
+
+  for (auto &ids : *input)
+  {
+    unsigned int detId = TotemRPDetId::rawToDecId(ids.detId());
+    unsigned int rpId = TotemRPDetId::rpOfDet(detId);
+    unsigned int plane = detId % 10;
+    bool uDir = TotemRPDetId::isStripsCoordinateUDirection(detId);
+
+    RPData &data = rpData[rpId];
+
+    for (auto &h : ids)
+    {
+      if (uDir)
+      {
+        auto &ods = data.hits_U.find_or_insert(ids.detId());
+        ods.push_back(h);
+        data.planeOccupancy_U[plane]++;
+      } else {
+        auto &ods = data.hits_V.find_or_insert(ids.detId());
+        ods.push_back(h);
+        data.planeOccupancy_V[plane]++;
+      }
+    }
+  }
+
+  // track recognition pot by pot
+  for (auto it : rpData)
+  {
+    unsigned int rpId = it.first;
+    RPData &data = it.second;
+
+    // merge default and exceptional settings (if available)
+    unsigned int minPlanesPerProjectionToFit_U = minPlanesPerProjectionToFit;
+    unsigned int minPlanesPerProjectionToFit_V = minPlanesPerProjectionToFit;
+    double threshold_U = threshold;
+    double threshold_V = threshold;
+
+    for (auto ps : exceptionalSettings)
+    {
+      unsigned int setId = ps.getParameter<unsigned int>("rpId");
```

Comment:
please move checks of PSet out of the way. There should be no need to check it every event.
Save a config variable instead.